server/src/test/java/com/octopus/teamcity/opentelemetry/server/TeamCityBuildListenerTest.java

```diff
@@ -0,0 +1,85 @@
+        final Map<String, String> HEADERS = new HashMap<>() {{
+            put("testHeader1", "testHeaderValue1");
+            put("testHeader2", "testHeaderValue2");
+        }};
+        GlobalOpenTelemetry.resetForTest();
+        this.otelHelper = new OTELHelper(HEADERS, ENDPOINT);
+        this.buildListener = new TeamCityBuildListener(buildServerListenerEventDispatcher, otelHelper);
+    }
+
+    @Test
+    void illegalStateExceptionWhenHeadersEmpty(@Mock EventDispatcher<BuildServerListener> buildServerListenerEventDispatcher) {
+        assertThrows(IllegalStateException.class,() -> new TeamCityBuildListener(buildServerListenerEventDispatcher));
+    }
+
+    @Test
+    void buildStartedAndOTELHelperIsReady() {
+        assertTrue(this.otelHelper.isReady());
+    }
+
+    @Test
+    void buildStartedWithNoParent() {
+        SRunningBuild build = mock(SRunningBuild.class, RETURNS_DEEP_STUBS);
+        // Stubbing this method to return the build if there is no parent, this is the behaviour of TeamCity
+        BuildPromotion[] buildPromotions = new BuildPromotion[]{build.getBuildPromotion()};
+        when(build.getBuildPromotion().findTops()).thenReturn(buildPromotions);
+        this.buildListener.buildStarted(build);
+        Span builtSpan = this.otelHelper.getSpan(String.valueOf(build.getBuildId()));
+        Span expectedSpan = this.otelHelper.createSpan(String.valueOf(build.getBuildId()), builtSpan);
+        assertEquals(expectedSpan, builtSpan);
+    }
+
+    @Test
+    void buildStartedWithParent() {
+        SRunningBuild build = mock(SRunningBuild.class, RETURNS_DEEP_STUBS);
+        SRunningBuild parentBuild = mock(SRunningBuild.class, RETURNS_DEEP_STUBS);
+        // Stubbing this method to return the build if there is no parent, this is the behaviour of TeamCity
+        BuildPromotion[] buildPromotions = new BuildPromotion[]{build.getBuildPromotion()};
+        when(build.getBuildPromotion().findTops()).thenReturn(buildPromotions);
+        this.buildListener.buildStarted(build);
+        Span parentSpan = this.otelHelper.getParentSpan(String.valueOf(parentBuild.getBuildId()));
+        Span builtSpan = this.otelHelper.getSpan(String.valueOf(build.getBuildId()));
+        Span expectedSpan = this.otelHelper.createSpan(String.valueOf(build.getBuildId()), parentSpan);
+        assertEquals(expectedSpan, builtSpan);
+    }
+
+    @Test
+    void buildFinishedOrInterrupted() {
+        SRunningBuild build = mock(SRunningBuild.class, RETURNS_DEEP_STUBS);
+        // Stubbing this method to return the build if there is no parent, this is the behaviour of TeamCity
+        BuildPromotion[] buildPromotions = new BuildPromotion[]{build.getBuildPromotion()};
+        when(build.getBuildPromotion().findTops()).thenReturn(buildPromotions);
+        this.buildListener.buildStarted(build);
+        this.buildListener.buildFinished(build);
+        assertNull(this.otelHelper.getSpan(String.valueOf(build.getBuildId())));
```

Comment:
Consider directly asserting on the map

Reply:
This is directly asserting on the spanMap as it is private to otelHelper and is avaiable by this getter